src/mos_editor.c

```diff
@@ -601,4 +576,87 @@ UINT24 mos_EDITLINE(char * buffer, int bufferLength, UINT8 flags) {
 	while (len-- > 0) putch(0x09);	// Then cursor right for the remainder
 
 	return keyr;					// Finally return the keycode
-}
\ No newline at end of file
+}
+
+void editHistoryInit() {
+	int i;
+	history_no = 0;
+	history_size = 0;
+
+	for (i = 0; i < cmd_historyDepth; i++) {
+		cmd_history[i] = NULL;
+	}
+}
+
+void editHistoryPush(char *buffer) {
+	int len = strlen(buffer);
+
+	if (len > 0) {		// If there is data in the buffer
+		char * newEntry = NULL;
+
+		// if the new entry is the same as the last entry, then don't save it
+		if (history_size > 0 && strcmp(buffer, cmd_history[history_size - 1]) == 0) {
+			return;
+		}
+
+		newEntry = umm_malloc(len + 1);
+		if (newEntry == NULL) {
+			// Memory allocation failed so we can't save history
+			return;
+		}
+		strcpy(newEntry, buffer);
+
+		// If we're at the end of the history, then we need to shift all our entries up by one
+		if (history_size == cmd_historyDepth) {
+			int i;
+			umm_free(cmd_history[0]);
+			for (i = 1; i < history_size; i++) {
+				cmd_history[i - 1] = cmd_history[i];
+			}
+			history_size--;
+		}
+		cmd_history[history_size++] = newEntry;
+		history_no = history_size;

```

Comment:
if we were using a linked list approach for the history this would be simplified to replace the current "head" of the history list with the new entry (which would point to the previous head), and optionally to trim the list

future ideas/plans for exposing history via system variables could be complicated by that tho, so for now we're using a simple array.  when system variables are added and if/when the history gets exposed this idea may be revisited